client_plugin/drivers/shared/shared_driver.go

```diff
@@ -23,24 +23,90 @@ package shared
+   dockerUSocket:       Unix socket on which Docker engine is listening
+   tradVolumePrefix:    Prefix for the names of internal volumes. These
+                        volumes server as backend stores for shared volumes.
+   sambaImageName:      Name of the docker hub image we pull as samba server
+   sambaUsername:       Default username for all accessing Samba servers
+   sambaPassword:       Default password for accessing all Samba servers
+
+   volStateCreating:    Shared volume is being created. Not ready to be mounted.
+   volStateReady:       Shared volume is ready to be mounted but
+                        no Samba service running right now.
+   volStateMounted:     Samba service already running. Volume mounted
+                        on at least one host VM.
+   volStateIntermediate: Metadata of shared volume is being changed.
+                         Dont proceed with your op if this is the status.
+*/
 const (
-	version       = "vSphere Shared Volume Driver v0.2"
-	apiVersion    = "v1.24" // docker engine 1.24 and above support this api version
-	dockerUSocket = "unix:///var/run/docker.sock"
+	version          = "vSphere Shared Volume Driver v0.2"
+	apiVersion       = "v1.24"
+	dockerUSocket    = "unix:///var/run/docker.sock"
+	tradVolumePrefix = "InternalTrad"
+	// TODO Replace with our own samba later
+	sambaImageName                 = "dperson/samba"
+	sambaUsername                  = "root"
+	sambaPassword                  = "badpass"
+	volStateCreating     volStatus = "Creating"
+	volStateReady        volStatus = "Ready"
+	volStateMounted      volStatus = "Mounted"
+	volStateIntermediate volStatus = "Intermediate"
 )
 
-// VolumeDriver - vsphere shared plugin volume driver struct
+/* VolumeMetadata structure contains all the
+   metadata about a volume that will be put in etcd
+
+   status:          What state is the shared volume currently in?
+   globalRefcount:  How many host VMs are accessing this volume?
+   port:            On which port is the Samba service listening?
+   serviceName:     What is the name of the Samba service for this volume?
+   username:
+   password:        Local Samba username and password
+	            Only default values for now, later can be used
+                    for multi tenancy.
+   clientList:      List of all host VMs using this shared volume
+*/
+type VolumeMetadata struct {
+	Status         volStatus `json:"-"` // Field won't be marshalled
+	GlobalRefcount int       `json:"-"` // Field won't be marshalled
+	Port           int       `json:"port"`
+	ServiceName    string    `json:"serviceName"`
+	Username       string    `json:"username"`
+	Password       string    `json:"password"`
+	ClientList     []string  `json:"clientList"`
+}
+
+/* VolumeDriver - vsphere shared plugin volume driver struct
+   dockerd:                 Client used for talking to Docker
+   internalVolumeDriver:    Name of the plugin used by shared volume
```

Comment:
We have a pre requisite for the user to install both the vDVS (or any other backing plugin) and the shared volume plugin on the system. This is something the user has to do.

But once this setup is correctly done, the shared volume plugin only talks to docker API over unix socket and mentions the "driver" it wants to use. The default is "vsphere" which refers to vDVS plugin.

It the same as when we do volume creation manually via CLI: docker volume create --driver vsphere vol1

Version information is not needed to make this API call, hence it is not required in the shared volume metadata.